### ISSUE TITLE
[Error [UnhandledRejection Error]: jpegsave: parameter in not set

### ISSUE THREAD

**Issue by gajop** [Jazcash/spring-map-parser#1]
Code:

```
	const destinationPath = path.join(springPlatform.writePath, command.destination);
	const parser = new MapParser({ verbose: true, mipmapSize: 4 });
	const map = await parser.parseMap(command.mapPath);

	if (destinationPath.endsWith('.jpg')) {
		await map.miniMap.jpeg({quality: 50 }).toFile(destinationPath);
	} else {
		await map.miniMap.toFile(destinationPath);
	}
```

Input: 

mapPath: `/home/gajop/Beyond All Reason/maps/dsdr_3.98.sd7`
destinationPath: `/home/gajop/Beyond All Reason/LuaMenu/Images/MinimapThumbnails/DSDR 3.98.jpg`

Output:

```
Parsing mapinfo.lua
Parsing .smf
Parsing .smt at mipmap size 4
Cleaning up temp dir: /tmp/dsdr_3.98
/home/gajop/Beyond All Reason/LuaMenu/Images/MinimapThumbnails/DSDR 3.98.jpg
[Error [UnhandledRejection Error]: jpegsave: parameter in not set
jpegsave: parameter filename not set
] {
  name: 'UnhandledRejection Error'
}
```

**Comment by gajop** [comment]
I also get a very similar error when using a .png as output

**Comment by gajop** [comment]
And the error isn't contained to minimaps only. I can't export textureMap either

**Comment by Jazcash** [comment]
Seems like a problem with libvips used in the [sharp](https://github.com/lovell/sharp) module. What version of node are you using? Latest 12.x.x and 14.x.x on both Windows and Linux seem ok for me. If it comes to it I'll ditch that module, it has large overhead anyway.

**Comment by gajop** [comment]
I'm on Linux with:
```
node --version
v12.8.1
```

Is there a way to get more information on what this failure relates to? Catching and printing the error.stack doesn't reveal much more info.

**Comment by Jazcash** [comment]
The error logs from sharp don't seem to be great. The problem could be passing an absolute path to `toFile` rather than a relative one. Could you try testing sharp on its own to see if it works at all in your environment?

`npm install sharp`

```ts
const sharp = require('sharp');

(async () => {
    await sharp("/home/some/absolute/path/file.png").toFile("/home/some/other/absolute/path/newfile.jpg");
})();
```

**Comment by gajop** [comment]
Well this works (generates a mini.jpg):

```js
it("minimap-fullpath-export", async () => {
    // expect.assertions(1);

    const mapPath = path.join(testMapsDir, "red_comet.sd7");

    const parser = new MapParser({ verbose: true, mipmapSize: 4, skipSmt: true });

    const map = await parser.parseMap(mapPath);

    await map.miniMap.jpeg({ quality: 50 }).toFile("/home/gajop/work/spring-map-parser/test/mini.jpg");
    expect(fs.existsSync("/home/gajop/work/spring-map-parser/test/mini.jpg")).toBe(true);
});
```

but like most other tests, it's still giving me `ENOTEMPTY: directory not empty, rmdir '/tmp/red_comet'`

This works, and doesn't give me any error, test or otherwise:

```js
it("minimap-fullpath-export", async () => {
    const sharp = require('sharp');

    (async () => {
        await sharp("/home/gajop/work/spring-map-parser/test/mini.jpg").toFile("/home/gajop/work/spring-map-parser/test/mini2.jpg");
    })();
});
```

My current guess is that the original error might be related to electron/sharp or javascript compatibility?

**Comment by gajop** [comment]
There's definitely something different between spring-launcher and the spring-map-parser.

This same code works if run as part of spring-map-parser tests, but doesn't when run as part of spring-launcher init:

```js
(async () => {
	const parser = new MapParser({ verbose: true, mipmapSize: 4, skipSmt: true });
	const map = await parser.parseMap('/home/gajop/Beyond All Reason/maps/dsdr_3.98.sd7');
	await map.miniMap.toFile('test.png');
})();
```

Porting Electron to a newer version (8.5.2 => 11.1.1) doesn't seem to fix the problem either.

**Comment by Jazcash** [comment]
[test.zip](https://github.com/Jazcash/spring-map-parser/files/5798113/test.zip)

Could you unzip this, `npm i` then `npm start` and see if that works in your environment?

**Comment by gajop** [comment]
@Jazcash 

No it doesn't. This is what I get:

```
npm start

> test@1.0.0 start /home/gajop/Downloads/test
> electron .

(node:45279) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
Extracting .sd7 to /tmp/bluebend-v01
Parsing .smd
Parsing .smf
Parsing .smt at mipmap size 4
Cleaning up temp dir: /tmp/bluebend-v01
(node:45279) UnhandledPromiseRejectionWarning: Error: jpegsave: parameter in not set
jpegsave: parameter filename not set

(node:45279) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:45279) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by gajop** [comment]
[test_gajop.zip](https://github.com/Jazcash/spring-map-parser/files/5815154/test_gajop.zip)
I did a further modification to it (including only sharp), and it gives me even weirder errors:

```
➜  test npm start

> test@1.0.0 start /home/gajop/Downloads/test
> electron .

Error: basic_string::_S_construct null not valid
(node:50048) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
Error: basic_string::_S_construct null not valid
```

I think I've hit https://github.com/lovell/sharp/issues/2356, and I'm looking to resolve it

**Comment by gajop** [comment]
@Jazcash After a bit more poking around, I've found that 0.26 and 0.27 versions of sharp don't work on my system, and that downgrading to 0.25 does (https://github.com/lovell/sharp/issues/2535). Interestingly enough neither of these versions gave me the kind of error I was reporting here originally - I suppose certain errors get lost between some of the abstraction layers.

I've found a hacky way to alter the sharp version used by spring-map-parser locally, and I've noticed that it finally works: I can actually get a minimap out of the bluebend-v01 you provided!

So you can you simply downgrade sharp to 0.25 for the time being?

**Comment by Jazcash** [comment]
Nice find. I've fixed sharp at v0.25.0 in spring-map-parser@1.2.0. Had a quick look through the [changelog ](https://sharp.pixelplumbing.com/changelog) but I don't think there's anything we'll miss.
